internal/envconfig/envconfig.go

```diff
@@ -12,168 +12,499 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package envconfig resolves configuration and secret values from environment
-// variables. This works by transforming the environment variables and then
-// invoking github.com/kelseyhightower/envconfig to map the environment
-// variables into a provided struct.
+// Package envconfig populates struct fields based on environment variable
+// values (or anything that responds to "Lookup"). Structs declare their
+// environment dependencies using the `env` tag with the key being the name of
+// the environment variable, case sensitive.
 //
-// If an environment variable begins with "secret://", the remaining string bits
-// are used to resolve the value in the provided SecretManager. For example:
+//     type MyStruct struct {
+//         A string `env:"A"` // resolves A to $A
+//         B string `env:"B,required"` // resolves B to $B, errors if $B is unset
+//         C string `env:"C,default=foo"` // resolves C to $C, defaults to "foo"
 //
-//     FOO=secret://foo/bar/baz => secretmanager.GetSecretValue("foo/bar/baz")
+//         D string `env:"D,required,default=foo"` // error, cannot be required and default
+//         E string `env:""` // error, must specify key
+//     }
 //
-// The environment variables are rewritten to be the secret value, but this is
-// only visible within the running process and any child processes.
+// All built-in types are supported except Func and Chan. If you need to define
+// a custom decoder, implement Decoder:
 //
-// If an environment variable secret ends with "?target=file" then the resulting
-// secret value is written to SECRETS_DIR and the environment variable is
-// updated to be the local path to that file.
+//     type MyStruct struct {
+//         field string
+//     }
+//
+//     func (v *MyStruct) EnvDecode(val string) error {
+//         v.field = fmt.Sprintf("PREFIX-%s", val)
+//         return nil
+//     }
+//
+// In the environment, slices are specified as comma-separated values:
+//
+//     export MYVAR="a,b,c,d" // []string{"a", "b", "c", "d"}
+//
+// In the environment, maps are specified as comma-separated key:value pairs:
+//
+//     export MYVAR="a:b,c:d" // map[string]string{"a":"b", "c":"d"}
+//
+// If you need to modify environment variable values before processing, you can
+// specify a custom mutator:
+//
+//     type Config struct {
+//         Password `env:"PASSWORD_SECRET"`
+//     }
+//
+//     func resolveSecretFunc(key, value string) (string, error) {
+//         if strings.HasSuffix(key, "_SECRET") {
+//             return secretmanager.Resolve(value)
+//         }
+//     }
+//
+//     var config Config
+//     ProcessWith(&config, new(OsLookuper), resolveSecretFunc)
+//
+// It is conceptually similar to kelseyhightower/envconfig, with the following
```

Comment:
Is it worth it or possible to import that and add the behavior changes on top, rather than reinvent from scratch? I don't know the landscape well enough; are there other standard config libraries that are worth mentioning (e.g. not using "foo" because xyz)?

Reply:
Ah, I see below we were actually straightup importing "github.com/kelseyhightower/envconfig" and now we've rolled our own. Maybe its worth a mention of other design options that were considered.

Reply:
I commented in the PR body about why we can't use it. It doesn't provide the extensibility hooks we need.

Reply:
I, too, would prefer to see a simpler approach.

we first did that by using kelseyhightower/envconfig to load the environment and then post-processing all the secret values.

I thought that's the direction we were headed back towards (at least that's the proposal in the issue)

Reply:
Per chat, there are challenges with kenvconfig that aren't easily patched or contributed upstream. I've pulled all the core logic into sethvargo/envconfig and cleaned this up.